File: python/array/134. Gas Station.py
'''
brute force: O(n^2)
Greedy
Time: O(n)
Space: O(1)
'''

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        if sum(gas) < sum(cost):  # no solution
            return -1
        
        total = 0
        res = 0  # start position
        for i in range(len(gas)):
            total += (gas[i] - cost[i])

            #  if total[0~i] < 0, and we know gas[0] - cost[0] >= 0, so total[1~i] always < 0
            if total < 0:
                res = i + 1 
                total = 0

        return res

File: javascript/array/134. Gas Station.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function(gas, cost) {
    const sumGas = gas.reduce((acc, val) => acc + val);
    const sumCost = cost.reduce((acc, val) => acc + val);
    if (sumGas < sumCost) return -1;

    let res = 0, total = 0;
    for (let i=0; i<gas.length; i++) {
        total += gas[i] - cost[i];

        if (total < 0) {
            res = i + 1;
            total = 0;
        }
    }
    return res;
}; 

// more suitable for javascript
var canCompleteCircuit = function(gas, cost) {
    let res = 0, curSum = 0, sum = 0;

    for (let i=0; i<gas.length; i++) {
        curSum += gas[i] - cost[i];
        sum += gas[i] - cost[i];
        if (curSum < 0) {
            res = i + 1;
            curSum = 0;
        }
    }
    if (sum < 0) return -1;

    return res;
}; 